packages/utils/crates/utils-types/src/sma.rs

```diff
@@ -0,0 +1,101 @@
+use ringbuffer::{AllocRingBuffer, RingBuffer};
+use std::fmt::{Display, Formatter};
+use std::ops::{Add, Div};
+
+/// Basic implementation of Simple Moving Average (SMA).
+/// The maximum window size is bound by 2^32 - 1.
+/// The second-most useful filter type, bested only by coffee filters.
+#[derive(Clone, Debug, PartialEq)]
+pub struct SMA<T>
+where
+    T: for<'a> Add<&'a T, Output = T> + Div<T, Output = T> + Clone + From<u32> + Display + Default,
+{
+    window: AllocRingBuffer<T>,
```

Comment:
For our implementation we don't need the values, let alone an entire buffer. We need only the average value + total count of samples constituting that average value (is the total value of samples needed at all?).

I wonder, whether this setup would be sufficient, since it is in constant time:
```rust
pub struct SMT<T, const N: usize>
{
    average: T,
    window_size: usize
}

pub fn add<T, const N: usize>(&mut self, value: T) {
    self.average = (self.average * (self.window_size-1) + new_value) / self.window_size
    self.window_size = N.min(self.window_size + 1)
}
```

The significance of earlier values would tail to 0 eventually, also, the constructor would have to assert the `N >= 1`.

Or maybe my fantasy is playing too much.

Reply:
We cannot have the window size as a compile time (const template) argument, as it needs to be configurable from the yaml. 

Next, we do need the number of samples added, which is used by the Promiscuous strategy to know when it actually has enough data. 

Also - I was thinking about the same constant time implementation, but this one has as a significant flaw - it accumulates floating point errors over time.

We can use a sum-tree method, which would be constant time for `get_average`, but logarithmic in `add_sample` and significantly more complex to implement. 
Given the fact that we much more often call `add_sample` than `get_average`, I stayed with this simple implementation.

Reply:
Ok, but implementation details of the strategu shiuld npt concern this windowed average object, the total sample count should live in the strategy, whereas this is a separate object.

Another question is, whether we care about any compounding floating errors. This is, after all, an already imprecise value  averaged.

Reply:
I believe we might overtime improve the underlying quality concept to be something more precise. And we can also swap the internal implementation to something else (single-sum or sum-tree SMA) at any time.

Looking at the Promiscuous strategy, it actually cares only if the window is not fully populated, so that can be improved.